### ISSUE TITLE
Pretty Printer

### ISSUE THREAD

**Issue by xurtis** [studley13/sexp#3]
Currently, all output is through display and printed on a single line. It'd be good to have a pretty printer that nicely formats the output using the following rules:
* Variable indentation rules (n-spaces or tabs)
* Loose or tight bracing options
* S-expressions with no child lists are printed on one line with either *loose* or *tight* bracing
* S-expressions with child sub-lists:
  * *Sparse* formatting places the first brace on its own line. Each value is placed on its own line
    with indentation advanced beyond the brace.
  * *Condensed* format places the first brace on the same line as the first value. The first value
    is then indented beyond the brace to line up with the following arguments.
  * The last brace is always on its own line, unindented

***Condensed* format with *tight* bracing**
```
( (blah blah blah)
  ( blah
    (blah blah blah)
  )
  blah
)
```
***Sparse* formatting with *loose* bracing**
```
( 
  ( blah blah blah )
  ( 
    blah
    ( blah blah blah )
  )
  blah
)
```

**Comment by jashank** [comment]
There is much prior art in s-expression pretty printing, though most of it focuses specifically on rules for Lisp-like languages. If that's your target here, you probably want to implement (something like) the [X3J13 pretty-printing style][x3j13-pp] or mimic the [Guile][guile-pp] or [Racket][racket-pp] pretty printers.

[x3j13-pp]: http://www.lispworks.com/documentation/HyperSpec/Issues/iss270.htm
[guile-pp]: https://www.gnu.org/software/guile/manual/html_node/Pretty-Printing.html
[racket-pp]: https://docs.racket-lang.org/reference/pretty-print.html

In summary, the former, tightly-braced style makes a good deal more sense as a starting point, but adding behaviours depending on symbols would be useful. Symbols should also always be tightly-braced before and after, as dangling parens or loose parens tend to look bad in the context of Lisps, so:

```
((blah blah blah)
  (blah
    (blah blah blah))
  blah)
```

**Comment by xurtis** [comment]
Let me propose a few rules.

# Classifying Values

First, we classify every value as either **code** or **data** in an implicit manner. 
* A list is considered **data** when:
  * it is explicitly marked as **data**, or
  * it begins with an explicit **data** value, or
  * it begins with a non-symbol value.
* Any list beginning with a symbol that is either explicitly marked as **code** or is not 
  marked as **data** is itself considered to be **code**.
* Symbols are considered **code** unless marked otherwise.
* All other values are considered **data**.

Secondly, we classify the **depth** of each list as such:
* `Int`, `Float`, and `String` all have a depth of **0**.
* The depth of a *cons pair* that is not a list is equal to the depth of its deepest member.
* The depth of a *list* is equal to the depth of its deepest member + 1

The hard part after this is using this information to know when to indent an break lines...

**Comment by fwouts** [comment]
Hi there 👋 I'm a few years late to the party, but was wondering if you intended to merge that `pretty-print` branch? I've been looking for a Rust sexp pretty-printer and this looks like exactly what I need :)
